File: src/main/java/sigmaone/industrialism/item/ItemWireSpool.java
package sigmaone.industrialism.item;

import net.minecraft.block.entity.BlockEntity;
import net.minecraft.item.Item;
import net.minecraft.item.ItemUsageContext;
import net.minecraft.text.LiteralText;
import net.minecraft.text.TranslatableText;
import net.minecraft.util.ActionResult;
import sigmaone.industrialism.block.wiring.BlockEntityWireNode;

public class ItemWireSpool extends Item {
    public BlockEntity linkA;

    public ItemWireSpool(Settings settings) {
        super(settings);
        this.linkA = null;
    }

    public void setLinkA(BlockEntity entity) {
        this.linkA = entity;
    }

    @Override
    public ActionResult useOnBlock(ItemUsageContext context) {
        if (!context.getWorld().isClient) {
            if (context.getWorld().getBlockEntity(context.getBlockPos()) instanceof BlockEntityWireNode) {
                if (this.linkA == null) {
                    this.setLinkA(context.getWorld().getBlockEntity(context.getBlockPos()));
                    context.getPlayer().sendMessage(new TranslatableText("item.industrialism.wire.popup.source", "("+this.linkA.getPos().toShortString()+")"), true);
                    return ActionResult.SUCCESS;
                }
                else if (this.linkA.getPos() == context.getBlockPos()) {
                    context.getPlayer().sendMessage(new TranslatableText("item.industrialism.wire.popup.failed.self"), true);
                    return ActionResult.PASS;
                }
                else if (this.linkA.getPos() != context.getBlockPos()) {
                    if (!((BlockEntityWireNode) this.linkA).getConnections().contains(context.getBlockPos()) && !((BlockEntityWireNode) context.getWorld().getBlockEntity(context.getBlockPos())).getConnections().contains(this.linkA.getPos())) {
                        if (!((BlockEntityWireNode) this.linkA).addConnection(context.getBlockPos())) {
                            context.getPlayer().sendMessage(new TranslatableText("item.industrialism.wire.popup.failed"), true);
                            this.linkA = null;
                            return ActionResult.PASS;
                        }
                        if (!((BlockEntityWireNode) context.getWorld().getBlockEntity(context.getBlockPos())).addConnection(this.linkA.getPos())) {
                            context.getPlayer().sendMessage(new TranslatableText("item.industrialism.wire.popup.failed"), true);
                            ((BlockEntityWireNode) this.linkA).removeConnection(context.getBlockPos());
                            this.linkA = null;
                            return ActionResult.PASS;
                        }
                        context.getPlayer().sendMessage(new TranslatableText("item.industrialism.wire.popup.linked", "("+this.linkA.getPos().toShortString()+")", "("+context.getBlockPos().toShortString()+")"), true);
                        this.linkA = null;
                        return ActionResult.SUCCESS;
                    }
                }
            }
        }
        return ActionResult.PASS;
    }
}

File: src/main/kotlin/sigmaone/industrialism/item/ItemWireSpool.kt
package sigmaone.industrialism.item

import net.minecraft.block.entity.BlockEntity
import net.minecraft.item.Item
import net.minecraft.item.ItemUsageContext
import net.minecraft.state.property.Properties
import net.minecraft.text.TranslatableText
import net.minecraft.util.ActionResult
import sigmaone.industrialism.block.wiring.BlockEntityWireConnectorT0
import sigmaone.industrialism.component.wiring.IComponentWireNode
import kotlin.math.pow
import kotlin.math.sqrt

class ItemWireSpool(
        settings: Settings?,
        val maxLength: Int,
        val thickness: Float,
        val colour: IntArray,
        val isInfinite: Boolean = false
    ) : Item(settings) {
    var linkA: BlockEntity? = null

    override fun useOnBlock(context: ItemUsageContext): ActionResult {
        if (!context.world.isClient) {
            val entity = context.world.getBlockEntity(context.blockPos)
            if (entity is IComponentWireNode) {
                if (this !in entity.componentWireNode.wireTypes) {
                    context.player!!.sendMessage(
                        TranslatableText("popup.industrialism.wire.failed.type"),
                        true
                    )
                    return ActionResult.PASS
                }

                if (linkA == null) {
                    linkA = entity
                    context.player!!.sendMessage(
                        TranslatableText(
                            "popup.industrialism.wire.source", "(" + linkA!!.pos.toShortString() + ")"
                        ),
                        true
                    )
                    return ActionResult.SUCCESS
                }

                else if (linkA!!.pos == context.blockPos) {
                    context.player!!.sendMessage(
                        TranslatableText("popup.industrialism.wire.failed.self"),
                        true
                    )
                    linkA = null
                    return ActionResult.PASS

                }
                else if (linkA!!.pos !== context.blockPos) {
                    val linkEntity = linkA!!
                    if (linkEntity is IComponentWireNode && linkA != null) {
                        if (!linkEntity.componentWireNode.connections.contains(context.blockPos)
                         && !linkEntity.componentWireNode.connections.contains((linkA as BlockEntityWireConnectorT0).pos)
                        ) {
                            val diff = sqrt(
                                (context.blockPos.y - linkEntity.pos.y).toFloat().pow(2)
                                + (
                                        (context.blockPos.x - linkEntity.pos.x).toFloat().pow(2)
                                        + (context.blockPos.z - linkEntity.pos.z).toFloat().pow(2)
                                  )
                            )
                            if (diff > maxLength) {
                                context.player!!.sendMessage(TranslatableText("popup.industrialism.wire.failed.toofar"), true)
                                linkA = null
                                return ActionResult.PASS
                            }
                            if (!linkEntity.componentWireNode.addConnection(
                                        context.blockPos,
                                        context.world.getBlockState(context.blockPos).get(Properties.FACING),
                                        entity.componentWireNode.height,
                                        this
                                    )
                            ) {
                                context.player!!.sendMessage(TranslatableText("popup.industrialism.wire.failed.generic"), true)
                                linkA = null
                                return ActionResult.PASS
                            }
                            if (!entity.componentWireNode.addConnection(
                                        linkEntity.pos,
                                        context.world.getBlockState(linkEntity.pos).get(Properties.FACING),
                                        linkEntity.componentWireNode.height,
                                        this
                                    )
                            ) {
                                context.player!!.sendMessage(TranslatableText("popup.industrialism.wire.failed.generic"), true)
                                linkEntity.componentWireNode.removeConnection(context.blockPos)
                                linkA = null
                                return ActionResult.PASS
                            }

                            context.player!!.sendMessage(
                                TranslatableText(
                                    "popup.industrialism.wire.linked", "(" + (linkA as BlockEntityWireConnectorT0).pos.toShortString() + ")",
                                    "(" + context.blockPos.toShortString() + ")"
                                ),
                                true
                            )

                            linkA = null
                            if (!isInfinite && !context.player!!.isCreative) {
                                context.stack.decrement(1)
                            }
                            return ActionResult.SUCCESS
                        }
                        else {
                            context.player!!.sendMessage(TranslatableText("popup.industrialism.wire.failed.exists"), true)
                            linkA = null
                            return ActionResult.PASS
                        }
                    }
                }
            }
        }
        return ActionResult.PASS
    }
}